NoClassDefFoundError in integration test of infux-db2

Even if the test passes, we can see the following error in the log of the build:
```
2024-07-19T11:08:28.1765553Z io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.lang.NoClassDefFoundError: io/reactivex/rxjava3/exceptions/Exceptions
2024-07-19T11:08:28.1768669Z 	at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:372)
2024-07-19T11:08:28.1769804Z 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:68)
2024-07-19T11:08:28.1771399Z 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
2024-07-19T11:08:28.1772700Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2024-07-19T11:08:28.1774027Z 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
2024-07-19T11:08:28.1775023Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2024-07-19T11:08:28.1776144Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2024-07-19T11:08:28.1776764Z 	at java.base/java.lang.Thread.run(Thread.java:840)
2024-07-19T11:08:28.1777517Z Caused by: java.lang.NoClassDefFoundError: io/reactivex/rxjava3/exceptions/Exceptions
2024-07-19T11:08:28.1778401Z 	at io.reactivex.rxjava3.internal.subscribers.BasicFuseableSubscriber.fail(BasicFuseableSubscriber.java:109)
2024-07-19T11:08:28.1779406Z 	at io.reactivex.rxjava3.internal.operators.flowable.FlowableMap$MapSubscriber.onNext(FlowableMap.java:66)
2024-07-19T11:08:28.1780329Z 	at io.reactivex.rxjava3.subscribers.SerializedSubscriber.onNext(SerializedSubscriber.java:103)
2024-07-19T11:08:28.1781569Z 	at com.influxdb.client.internal.flowable.FlowableBufferTimedFlushable$BufferExactBoundedSubscriber.accept(FlowableBufferTimedFlushable.java:227)
2024-07-19T11:08:28.1782925Z 	at com.influxdb.client.internal.flowable.FlowableBufferTimedFlushable$BufferExactBoundedSubscriber.accept(FlowableBufferTimedFlushable.java:85)
2024-07-19T11:08:28.1784188Z 	at io.reactivex.rxjava3.internal.subscribers.QueueDrainSubscriber.fastPathOrderedEmitMax(QueueDrainSubscriber.java:107)
2024-07-19T11:08:28.1785444Z 	at com.influxdb.client.internal.flowable.FlowableBufferTimedFlushable$BufferExactBoundedSubscriber.run(FlowableBufferTimedFlushable.java:281)
2024-07-19T11:08:28.1786474Z 	at io.reactivex.rxjava3.core.Scheduler$Worker$PeriodicTask.run(Scheduler.java:543)
2024-07-19T11:08:28.1787269Z 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
2024-07-19T11:08:28.1787835Z 	... 6 more
2024-07-19T11:08:28.1788884Z Caused by: java.lang.ClassNotFoundException: Unable to load class 'io.reactivex.rxjava3.exceptions.Exceptions' because the bundle wiring for io.reactivex.rxjava3.rxjava is no longer valid.
2024-07-19T11:08:28.1790221Z 	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1542)
2024-07-19T11:08:28.1791123Z 	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:79)
2024-07-19T11:08:28.1791963Z 	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1976)
2024-07-19T11:08:28.1792715Z 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
2024-07-19T11:08:28.1793148Z 	... 15 more
2024-07-19T11:08:28.1796248Z ##[error]Exception in thread "RxNewThreadScheduler-1" io.reactivex.rxjava3.exceptions.UndeliverableException: The exception could not be delivered to the consumer because it has already canceled/disposed the flow or the exception has nowhere to go to begin with. Further reading: https://github.com/ReactiveX/RxJava/wiki/What's-different-in-2.0#error-handling | java.lang.NoClassDefFoundError: io/reactivex/rxjava3/exceptions/Exceptions
2024-07-19T11:08:28.1800704Z 	at io.reactivex.rxjava3.plugins.RxJavaPlugins.onError(RxJavaPlugins.java:372)
2024-07-19T11:08:28.1801637Z 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:68)
2024-07-19T11:08:28.1802528Z 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.call(ScheduledRunnable.java:56)
2024-07-19T11:08:28.1803280Z 	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
2024-07-19T11:08:28.1804166Z 	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:304)
2024-07-19T11:08:28.1805162Z 	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
2024-07-19T11:08:28.1806006Z 	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
2024-07-19T11:08:28.1806632Z 	at java.base/java.lang.Thread.run(Thread.java:840)
2024-07-19T11:08:28.1807214Z Caused by: java.lang.NoClassDefFoundError: io/reactivex/rxjava3/exceptions/Exceptions
2024-07-19T11:08:28.1808113Z 	at io.reactivex.rxjava3.internal.subscribers.BasicFuseableSubscriber.fail(BasicFuseableSubscriber.java:109)
2024-07-19T11:08:28.1809118Z 	at io.reactivex.rxjava3.internal.operators.flowable.FlowableMap$MapSubscriber.onNext(FlowableMap.java:66)
2024-07-19T11:08:28.1810050Z 	at io.reactivex.rxjava3.subscribers.SerializedSubscriber.onNext(SerializedSubscriber.java:103)
2024-07-19T11:08:28.1811258Z 	at com.influxdb.client.internal.flowable.FlowableBufferTimedFlushable$BufferExactBoundedSubscriber.accept(FlowableBufferTimedFlushable.java:227)
2024-07-19T11:08:28.1812619Z 	at com.influxdb.client.internal.flowable.FlowableBufferTimedFlushable$BufferExactBoundedSubscriber.accept(FlowableBufferTimedFlushable.java:85)
2024-07-19T11:08:28.1813889Z 	at io.reactivex.rxjava3.internal.subscribers.QueueDrainSubscriber.fastPathOrderedEmitMax(QueueDrainSubscriber.java:107)
2024-07-19T11:08:28.1815410Z 	at com.influxdb.client.internal.flowable.FlowableBufferTimedFlushable$BufferExactBoundedSubscriber.run(FlowableBufferTimedFlushable.java:281)
2024-07-19T11:08:28.1816478Z 	at io.reactivex.rxjava3.core.Scheduler$Worker$PeriodicTask.run(Scheduler.java:543)
2024-07-19T11:08:28.1817289Z 	at io.reactivex.rxjava3.internal.schedulers.ScheduledRunnable.run(ScheduledRunnable.java:65)
2024-07-19T11:08:28.1817871Z 	... 6 more
2024-07-19T11:08:28.1818914Z Caused by: java.lang.ClassNotFoundException: Unable to load class 'io.reactivex.rxjava3.exceptions.Exceptions' because the bundle wiring for io.reactivex.rxjava3.rxjava is no longer valid.
2024-07-19T11:08:28.1820254Z 	at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1542)
2024-07-19T11:08:28.1821171Z 	at org.apache.felix.framework.BundleWiringImpl.access$300(BundleWiringImpl.java:79)
2024-07-19T11:08:28.1822014Z 	at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:1976)
2024-07-19T11:08:28.1822776Z 	at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:525)
2024-07-19T11:08:28.1823214Z 	... 15 more
```

--- a/tests/camel-integration-test/src/main/java/org/apache/karaf/camel/itests/CamelRouteSupplier.java
+++ b/tests/camel-integration-test/src/main/java/org/apache/karaf/camel/itests/CamelRouteSupplier.java
@@ -23,4 +23,11 @@ public interface CamelRouteSupplier {
      * @param builder the Camel route builder
      */
     void createRoutes(RouteBuilder builder);
+
+    /**
+     * Cleans up the Camel context before removing the routes.
+     */
+    default void cleanUp(CamelContext camelContext) {
+        // Do nothing by default
+    }
 }

--- a/tests/camel-integration-test/src/main/java/org/apache/karaf/camel/itests/CamelSuppliedRouteLauncher.java
+++ b/tests/camel-integration-test/src/main/java/org/apache/karaf/camel/itests/CamelSuppliedRouteLauncher.java
@@ -124,6 +124,7 @@ public class CamelSuppliedRouteLauncher extends AbstractCamelRouteLauncher imple
                 return;
             }
             camelContext.removeRouteDefinitions(routeDefinitions);
+            supplier.cleanUp(camelContext);
             LOG.info("Route(s) removed from CamelRouteSupplier service: {}", supplier.getClass().getName());
         } catch (Exception e) {
             throw new RuntimeException(e);

--- a/tests/features/camel-arangodb/src/main/java/org/apache/karaf/camel/test/CamelArangodbRouteSupplier.java
+++ b/tests/features/camel-arangodb/src/main/java/org/apache/karaf/camel/test/CamelArangodbRouteSupplier.java
@@ -42,7 +42,7 @@ public class CamelArangodbRouteSupplier extends AbstractCamelSingleFeatureResult
 
     @Override
     public void configure(CamelContext camelContext) {
-        final int arangoPort = Integer.parseInt(System.getProperty("arango.port"));
+        final int arangoPort = Integer.getInteger("arango.port");
         ArangoDB arangoDb = new ArangoDB.Builder().host("localhost", arangoPort).build();
 
         arangoDb.createDatabase(DATABASE_NAME);
@@ -54,7 +54,14 @@ public class CamelArangodbRouteSupplier extends AbstractCamelSingleFeatureResult
         arangoDbComponent.getConfiguration().setHost("localhost");
         arangoDbComponent.getConfiguration().setPort(arangoPort);
 
-        camelContext.addComponent("arangodb",arangoDbComponent);
+        camelContext.addComponent("arangodb", arangoDbComponent);
+    }
+
+    @Override
+    public void cleanUp(CamelContext camelContext) {
+        if (camelContext.hasComponent("arangodb") instanceof ArangoDbComponent arangoDbComponent) {
+            arangoDbComponent.getArangoDB().shutdown();
+        }
     }
 
     @Override

--- a/tests/features/camel-as2/src/main/java/org/apache/karaf/camel/test/CamelAs2RouteSupplier.java
+++ b/tests/features/camel-as2/src/main/java/org/apache/karaf/camel/test/CamelAs2RouteSupplier.java
@@ -75,7 +75,7 @@ public class CamelAs2RouteSupplier extends AbstractCamelSingleFeatureResultMockB
 
     @Override
     public void configure(CamelContext context) {
-        final int port = Integer.parseInt(System.getProperty("as2.port"));
+        final int port = Integer.getInteger("as2.port");
         final AS2Configuration configuration = new AS2Configuration();
         configuration.setTargetHostname("localhost");
         configuration.setTargetPortNumber(port);

--- a/tests/features/camel-dns/pom.xml
+++ b/tests/features/camel-dns/pom.xml
@@ -35,7 +35,6 @@
             <groupId>dnsjava</groupId>
             <artifactId>dnsjava</artifactId>
             <version>${dnsjava-version}</version>
-            <scope>test</scope>
         </dependency>
     </dependencies>
 

--- a/tests/features/camel-dns/src/main/java/org/apache/karaf/camel/test/CamelDnsRouteSupplier.java
+++ b/tests/features/camel-dns/src/main/java/org/apache/karaf/camel/test/CamelDnsRouteSupplier.java
@@ -15,11 +15,13 @@
  */
 package org.apache.karaf.camel.test;
 
+import org.apache.camel.CamelContext;
 import org.apache.camel.builder.RouteBuilder;
 import org.apache.camel.model.RouteDefinition;
 import org.apache.karaf.camel.itests.AbstractCamelSingleFeatureResultMockBasedRouteSupplier;
 import org.apache.karaf.camel.itests.CamelRouteSupplier;
 import org.osgi.service.component.annotations.Component;
+import org.xbill.DNS.NioClient;
 
 @Component(
         name = "karaf-camel-dns-test",
@@ -33,6 +35,12 @@ public class CamelDnsRouteSupplier extends AbstractCamelSingleFeatureResultMockB
     @Override
     protected boolean consumerEnabled() {
         return false;
+    }
+
+    @Override
+    public void cleanUp(CamelContext camelContext) {
+        // Close it explicitly to prevent NoClassDefFoundError after the test
+        NioClient.close();
     }
 
     @Override

--- a/tests/features/camel-ehcache/src/main/java/org/apache/karaf/camel/test/CamelEhcacheRouteSupplier.java
+++ b/tests/features/camel-ehcache/src/main/java/org/apache/karaf/camel/test/CamelEhcacheRouteSupplier.java
@@ -51,7 +51,16 @@ public class CamelEhcacheRouteSupplier extends AbstractCamelSingleFeatureResultM
                 )
                 .build(true);
 
-        camelContext.getRegistry().bind("cacheManager",cacheManager);
+        camelContext.getRegistry().bind("cacheManager", cacheManager);
+    }
+
+    @Override
+    public void cleanUp(CamelContext camelContext) {
+        CacheManager cacheManager = camelContext.getRegistry().lookupByNameAndType("cacheManager", CacheManager.class);
+        if (cacheManager == null) {
+            return;
+        }
+        cacheManager.close();
     }
 
     @Override

--- a/tests/features/camel-elasticsearch/src/main/java/org/apache/karaf/camel/test/CamelElasticsearchRouteSupplier.java
+++ b/tests/features/camel-elasticsearch/src/main/java/org/apache/karaf/camel/test/CamelElasticsearchRouteSupplier.java
@@ -39,7 +39,7 @@ public class CamelElasticsearchRouteSupplier extends AbstractCamelSingleFeatureR
         elasticsearchComponent.setPassword(System.getProperty("elasticsearch.password"));
         elasticsearchComponent.setCertificatePath("file:%s".formatted(System.getProperty("elasticsearch.cafile")));
 
-        camelContext.addComponent("elasticsearch",elasticsearchComponent);
+        camelContext.addComponent("elasticsearch", elasticsearchComponent);
     }
 
     @Override

--- a/tests/features/camel-hazelcast/src/main/java/org/apache/karaf/camel/test/CamelHazelcastRouteSupplier.java
+++ b/tests/features/camel-hazelcast/src/main/java/org/apache/karaf/camel/test/CamelHazelcastRouteSupplier.java
@@ -36,10 +36,18 @@ public class CamelHazelcastRouteSupplier extends AbstractCamelSingleFeatureResul
 
     @Override
     public void configure(CamelContext camelContext) {
-        ClusterProperty.SHUTDOWNHOOK_ENABLED.setSystemProperty("false");
         Config hazelcastConfig = new Config();
         HazelcastInstance hazelcastInstance = Hazelcast.newHazelcastInstance(hazelcastConfig);
         camelContext.getRegistry().bind("hzInstance", hazelcastInstance);
+    }
+
+    @Override
+    public void cleanUp(CamelContext camelContext) {
+        HazelcastInstance hazelcastInstance = camelContext.getRegistry().lookupByNameAndType("hzInstance", HazelcastInstance.class);
+        if (hazelcastInstance == null) {
+            return;
+        }
+        hazelcastInstance.shutdown();
     }
 
     @Override

--- a/tests/features/camel-influxdb2/src/main/java/org/apache/karaf/camel/test/CamelInfluxdb2RouteSupplier.java
+++ b/tests/features/camel-influxdb2/src/main/java/org/apache/karaf/camel/test/CamelInfluxdb2RouteSupplier.java
@@ -38,11 +38,20 @@ public class CamelInfluxdb2RouteSupplier extends AbstractCamelSingleFeatureResul
 
     @Override
     public void configure(CamelContext camelContext) {
-        final int influxdb2Port = Integer.parseInt(System.getProperty("influxdb2.port"));
-        InfluxDBClient myInfluxDBClient = InfluxDBClientFactory.create("http://localhost:%s".formatted(influxdb2Port),
+        final int influxdb2Port = Integer.getInteger("influxdb2.port");
+        InfluxDBClient influxDBClient = InfluxDBClientFactory.create("http://localhost:%s".formatted(influxdb2Port),
                 System.getProperty("influxdb2.admin.token").toCharArray(), ORG, BUCKET);
 
-        camelContext.getRegistry().bind("myDbClient", myInfluxDBClient);
+        camelContext.getRegistry().bind("myDbClient", influxDBClient);
+    }
+
+    @Override
+    public void cleanUp(CamelContext camelContext) {
+        InfluxDBClient influxDBClient = camelContext.getRegistry().lookupByNameAndType("myDbClient", InfluxDBClient.class);
+        if (influxDBClient == null) {
+            return;
+        }
+        influxDBClient.close();
     }
 
     @Override

--- a/tests/features/camel-netty-http/src/main/java/org/apache/karaf/camel/test/CamelRestNettyHttpRouteSupplier.java
+++ b/tests/features/camel-netty-http/src/main/java/org/apache/karaf/camel/test/CamelRestNettyHttpRouteSupplier.java
@@ -34,7 +34,7 @@ public class CamelRestNettyHttpRouteSupplier extends AbstractCamelSingleFeatureR
         RestConfiguration config = new RestConfiguration();
         config.setComponent("netty-http");
         config.setHost("127.0.0.1");
-        config.setPort(Integer.parseInt(System.getProperty(REST_PORT)));
+        config.setPort(Integer.getInteger(REST_PORT));
         camelContext.setRestConfiguration(config);
     }
 

